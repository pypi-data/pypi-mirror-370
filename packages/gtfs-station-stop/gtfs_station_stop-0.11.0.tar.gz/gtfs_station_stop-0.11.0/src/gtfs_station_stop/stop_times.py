"""Stop Times Dataset."""

import os
from dataclasses import dataclass, field
from enum import Enum
from typing import Self

from .helpers import get_as_number
from .static_dataset import GtfsStaticDataset


class PickupType(Enum):
    """
    Pickup Type.
    see https://gtfs.org/documentation/schedule/reference/#stop_timestxt
    """

    REGULARLY_SCHEDULED = 0
    NO_PICKUP_AVAILABLE = 1
    PHONE_AGENCY_FOR_PICKUP = 2
    COORDINATE_WITH_DRIVER_FOR_PICKUP = 3


class DropOffType(Enum):
    """
    Drop-Off Type.
    see https://gtfs.org/documentation/schedule/reference/#stop_timestxt
    """

    REGULARLY_SCHEDULED = 0
    NO_DROP_OFF_AVAILABLE = 1
    PHONE_AGENCY_FOR_DROP_OFF = 2
    COORDINATE_WITH_DRIVER_FOR_DROP_OFF = 3


class TimePoint(Enum):
    """
    Timepoint
    see https://gtfs.org/documentation/schedule/reference/#stop_timestxt
    """

    APPROXIMATE = 0
    EXACT = 1


@dataclass(order=True)
class GtfsArrivalDepartureTime:
    """GTFS Specific Arrival or Departure Times from stoptimes.txt"""

    hour: int
    minute: int
    second: int

    @staticmethod
    def strptime(time_str: str) -> Self:
        """
        Generate the dataclass from a string in a dataset row.
        Times are permitted to exceed 24 hours for datasets after midnight, so the
        built-in method datetime.strptime cannot be used.
        see https://gtfs.org/documentation/schedule/reference/#stop_timestxt
        arrival_time and departure_time
        """
        return GtfsArrivalDepartureTime(*[int(x) for x in time_str.split(":")])


@dataclass
class StopTime:
    """Stop Time."""

    def __init__(self, stop_times_data_dict: dict) -> None:
        self.trip_id = stop_times_data_dict["trip_id"]

        self.arrival_time: GtfsArrivalDepartureTime | None = None
        if (arrival_time_str := stop_times_data_dict.get("arrival_time")) not in {
            None,
            "",
        }:
            self.arrival_time = GtfsArrivalDepartureTime.strptime(arrival_time_str)

        self.departure_time: GtfsArrivalDepartureTime | None = None
        if (departure_time_str := stop_times_data_dict.get("departure_time")) not in {
            None,
            "",
        }:
            self.departure_time = GtfsArrivalDepartureTime.strptime(departure_time_str)

        self.stop_id = stop_times_data_dict.get("stop_id")
        self.location_group_id = stop_times_data_dict.get("location_group_id")
        self.location_id = stop_times_data_dict.get("location_id")

        self.stop_sequence = int(stop_times_data_dict["stop_sequence"])
        self.stop_headsign = stop_times_data_dict.get("stop_headsign", "")
        self.start_pickup_drop_off_window = stop_times_data_dict.get(
            "start_pickup_drop_off_window"
        )
        self.end_pickup_drop_off_window = stop_times_data_dict.get(
            "end_pickup_drop_off_window"
        )
        self.pickup_type = PickupType(
            get_as_number(stop_times_data_dict, "pickup_type", int)
        )
        self.drop_off_type = DropOffType(
            get_as_number(stop_times_data_dict, "drop_off_type", int)
        )

        self.continuous_pickup = PickupType(
            get_as_number(stop_times_data_dict, "continuous_pickup", int)
        )

        self.continuous_drop_off = DropOffType(
            get_as_number(stop_times_data_dict, "continuous_drop_off", int)
        )

        self.shape_dist_traveled = float(
            get_as_number(stop_times_data_dict, "shape_dist_traveled", float)
        )
        self.timepoint = TimePoint(
            get_as_number(stop_times_data_dict, "timepoint", int, 1)
        )


@dataclass
class StopTimesDataset(GtfsStaticDataset):
    """Dataset for Stop Times."""

    stop_times: dict[str, dict[int, StopTime]]
    stops_filter: set[str] = field(default_factory=set)  # only add stop_ids from here

    def __init__(self, *gtfs_files: os.PathLike, **kwargs):
        self.stop_times = {}
        self.stops_filter = set()
        super().__init__(*gtfs_files, **kwargs)

    def add_gtfs_data(self, zip_filelike) -> None:
        for line in self._get_gtfs_record_iter(zip_filelike, "stop_times.txt"):
            if line.get("stop_id") in (self.stops_filter or set()):
                stop_time = StopTime(line)
                self.stop_times.setdefault(stop_time.trip_id, {})[
                    stop_time.stop_sequence
                ] = stop_time

    def get(self, trip_id, stop_sequence, *, default: StopTime | None = None):
        """Get Stop Time from Dataset."""
        by_trip_id = self.stop_times.get(trip_id)
        if by_trip_id is not None:
            return by_trip_id.get(stop_sequence, default)
        return default
