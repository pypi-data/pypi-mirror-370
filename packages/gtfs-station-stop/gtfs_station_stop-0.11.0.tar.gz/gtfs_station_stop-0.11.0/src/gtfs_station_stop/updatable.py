"""Updatable"""


class Updatable:
    """Updatable Base class."""

    _last_updated: float | None = None

    @property
    def last_updated(self) -> None:
        """Time Last Updated."""
        return self._last_updated

    def begin_update(self, timestamp: float | None = None) -> None:
        """Prepare for update by Feed Subject"""
        raise NotImplementedError
