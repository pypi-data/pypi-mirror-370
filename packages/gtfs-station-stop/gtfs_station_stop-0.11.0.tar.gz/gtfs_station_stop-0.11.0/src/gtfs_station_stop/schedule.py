"""Schedule"""

import os
import shutil
import tempfile
from asyncio import TaskGroup
from dataclasses import dataclass, field
from io import BytesIO
from os import PathLike
from pathlib import Path
from tempfile import TemporaryDirectory
from zipfile import ZipFile

import aiofiles
from aiohttp import ClientSession
from yarl import URL

from gtfs_station_stop.calendar import Calendar
from gtfs_station_stop.route_info import RouteInfo, RouteInfoDataset
from gtfs_station_stop.static_dataset import (
    GtfsStaticDataset,
    async_factory,
)
from gtfs_station_stop.station_stop_info import StationStopInfo, StationStopInfoDataset
from gtfs_station_stop.stop_times import StopTimesDataset
from gtfs_station_stop.trip_info import TripInfo, TripInfoDataset

DEFAULT_CHUNK_SIZE = 65536


async def _get_nested_zip(
    target: Path, dest: PathLike, *, chunk_size: int = DEFAULT_CHUNK_SIZE
) -> list[Path]:
    dest = Path(dest)
    if not dest.is_dir():
        raise ValueError("Must pass a directory as the `dest` parameter")

    res = []
    with ZipFile(target, "r") as z:
        for file in [
            file
            for file in z.filelist
            if not file.is_dir() and file.filename.endswith(".zip")
        ]:
            dest_fullpath = dest / file.filename
            os.makedirs(dest_fullpath.parent, exist_ok=True)
            with z.open(file, "r") as z_nested:
                async with (
                    aiofiles.open(dest_fullpath, "wb") as f,
                ):
                    buf = bytearray(chunk_size)
                    while (n_bytes := z_nested.readinto(buf)) > 0:
                        await f.write(buf[:n_bytes])
            res.append(dest_fullpath)
    return res


@dataclass(kw_only=True)
class GtfsSchedule:
    """GTFS Schedule."""

    calendar: Calendar = field(default_factory=Calendar)
    station_stop_info_ds: StationStopInfoDataset = field(
        default_factory=StationStopInfoDataset
    )
    trip_info_ds: TripInfoDataset = field(default_factory=TripInfoDataset)
    route_info_ds: RouteInfoDataset = field(default_factory=RouteInfoDataset)
    stop_times_ds: StopTimesDataset = field(default_factory=StopTimesDataset)
    tmp_dir: TemporaryDirectory | None = None
    tmp_dir_path: Path | None = None
    resources: set[Path] = field(default_factory=set)

    async def async_build_schedule(
        self,
        *gtfs_urls: URL,
        session: ClientSession | None = None,
        **kwargs,
    ) -> None:
        """Update the schedule given a set of URLs"""
        close_session = False
        if session is None:
            session = ClientSession()
            close_session = True

        try:
            if self.tmp_dir is None:
                gtfs_tmp_dir = f"{tempfile.gettempdir()}/gtfs_station_stop"
                os.makedirs(gtfs_tmp_dir, exist_ok=True)
                self.tmp_dir = TemporaryDirectory(dir=gtfs_tmp_dir)
                self.tmp_dir_path = Path(self.tmp_dir.name)
            async with TaskGroup() as tg:
                for url in gtfs_urls:
                    hash_str = hex(hash(str(url)) & 0xFFFFFFFF)[2:]
                    path = self.tmp_dir_path / f"{hash_str}.zip"
                    tg.create_task(
                        self._async_download_to_file_and_add_data(
                            tg,
                            url,
                            path,
                            session=session,
                            **kwargs,
                        )
                    )
        finally:
            if close_session:
                await session.close()

    async def async_load_stop_times(self, stops_filter: set[str] | None = None) -> None:
        """
        Async load stop times in stop_times.txt
        This operation should be deayed from schedule building as it can
        be time consuming and is not needed for many datasets.
        """
        self.stop_times_ds.stops_filter = stops_filter or set()
        for resource in self.resources:
            in_mem: BytesIO | None = None
            async with aiofiles.open(resource, "rb") as f:
                in_mem = BytesIO(await f.read())
                self.stop_times_ds.add_gtfs_data(in_mem)

    def get_stop_info(self, stop_id: str) -> StationStopInfo | None:
        """Get stop info by ID."""
        return self.station_stop_info_ds.station_stop_infos.get(stop_id)

    def get_trip_headsign(self, trip_id: str) -> str:
        """Get Trip's Headsign."""
        trip_info: TripInfo = self.trip_info_ds.get_close_match(trip_id)
        if trip_info is not None:
            return trip_info.trip_headsign
        return ""

    def get_route_color(self, route_id: str) -> str:
        """Get Trip's Route Color."""
        route_info: RouteInfo = self.route_info_ds.get(route_id)
        if route_info is not None:
            return route_info.color
        return ""

    def get_route_text_color(self, route_id: str) -> str:
        """Get Trip's Route Text Color."""
        route_info: RouteInfo = self.route_info_ds.get(route_id)
        if route_info is not None:
            return route_info.text_color
        return ""

    def get_route_type(self, route_id: str) -> str:
        """Get Trip's Route Type."""
        route_info: RouteInfo = self.route_info_ds.get(route_id)
        if route_info is not None:
            return route_info.type.pretty_name()
        return ""

    def __del__(self) -> None:
        if self.tmp_dir_path is not None and self.tmp_dir_path.is_dir():
            shutil.rmtree(self.tmp_dir_path, ignore_errors=True)

    def _get_required_datasets(self) -> list[GtfsStaticDataset]:
        return [
            self.calendar,
            self.station_stop_info_ds,
            self.trip_info_ds,
            self.route_info_ds,
        ]

    async def _async_download_to_file_and_add_data(
        self,
        task_group: TaskGroup,
        url: URL,
        target: Path,
        *,
        session: ClientSession | None = None,
        chunk_size: int = DEFAULT_CHUNK_SIZE,
        **kwargs,
    ) -> None:
        async with (
            session.get(url, **kwargs) as resp,
            aiofiles.open(target, "wb") as tmp_f,
        ):
            async for chunk in resp.content.iter_chunked(chunk_size):
                await tmp_f.write(chunk)

        self.resources.add(target)
        nested_dest = target.parent
        os.makedirs(nested_dest, exist_ok=True)
        nested_resources = [Path(s) for s in await _get_nested_zip(target, nested_dest)]
        self.resources.update(nested_resources)
        for resource in [target] + nested_resources:
            in_mem: BytesIO | None = None
            async with aiofiles.open(resource, "rb") as f:
                in_mem = BytesIO(await f.read())
            for ds in self._get_required_datasets():
                task_group.create_task(async_factory(ds, in_mem))


async def async_build_schedule(
    *gtfs_urls: os.PathLike,
    session: ClientSession | None = None,
    **kwargs,
) -> GtfsSchedule:
    """Build a schedule dataclass."""

    close_session: bool = False
    if session is None:
        session = ClientSession()
        close_session = True

    try:
        schedule = GtfsSchedule()
        await schedule.async_build_schedule(*gtfs_urls, session=session, **kwargs)
    finally:
        if close_session:
            await session.close()
    return schedule
