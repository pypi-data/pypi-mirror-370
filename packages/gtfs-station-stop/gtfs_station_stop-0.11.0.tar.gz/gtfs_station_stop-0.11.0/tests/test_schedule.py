"""Test Schedule."""

from dataclasses import asdict

import syrupy.filters
from aiohttp import ClientSession

from gtfs_station_stop.route_info import RouteInfoDataset
from gtfs_station_stop.schedule import (
    GtfsSchedule,
    async_build_schedule,
)
from gtfs_station_stop.stop_times import StopTimesDataset

schedule_filter = syrupy.filters.props("tmp_dir", "tmp_dir_path", "resources")


async def test_async_build_schedule(mock_feed_server, snapshot):
    async with ClientSession() as session:
        schedule: GtfsSchedule = await async_build_schedule(
            *[
                url
                for url in mock_feed_server.static_urls
                if url.endswith("gtfs_static.zip")
            ],
            session=session,
        )
    assert snapshot(exclude=schedule_filter) == asdict(schedule)
    assert isinstance(schedule.stop_times_ds, StopTimesDataset)


async def test_async_build_schedule_add_data_later(mock_feed_server, snapshot):
    schedule: GtfsSchedule = await async_build_schedule(
        *[
            url
            for url in mock_feed_server.static_urls
            if url.endswith("gtfs_static.zip")
        ]
    )
    orig_data = asdict(schedule)

    await schedule.async_build_schedule(
        *[
            url
            for url in mock_feed_server.static_urls
            if url.endswith("gtfs_static_supl.zip")
        ]
    )
    await schedule.async_load_stop_times()

    assert orig_data != asdict(schedule)
    assert snapshot(exclude=schedule_filter) == asdict(schedule)
    assert isinstance(schedule.stop_times_ds, StopTimesDataset)
    assert isinstance(schedule.route_info_ds, RouteInfoDataset)


async def test_stop_time_filtering(mock_feed_server, snapshot):
    schedule: GtfsSchedule = await async_build_schedule(
        *[
            url
            for url in mock_feed_server.static_urls
            if url.endswith("gtfs_static.zip")
        ]
    )
    await schedule.async_load_stop_times({"101N"})

    assert snapshot(
        exclude=syrupy.filters.props("tmp_dir", "tmp_dir_path", "resources")
    ) == asdict(schedule)

    assert schedule.stop_times_ds.get("STOP_TIME_TRIP", 1) is not None
