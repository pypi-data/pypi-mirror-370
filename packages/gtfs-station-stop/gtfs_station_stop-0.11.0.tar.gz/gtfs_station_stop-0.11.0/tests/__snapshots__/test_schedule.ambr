# serializer version: 1
# name: test_async_build_schedule
  dict({
    'calendar': dict({
      'services': dict({
        'NoFridayService': Service(
          added_exceptions=set({
          }),
          end=datetime.date(1, 1, 1),
          removed_exceptions=set({
            datetime.date(2024, 3, 1),
          }),
          service_days=ServiceDays(
            friday=False,
            monday=False,
            saturday=False,
            sunday=False,
            thursday=False,
            tuesday=False,
            wednesday=False,
          ),
          service_id='NoFridayService',
          start=datetime.date(1, 1, 1),
        ),
        'PiDaySpecial': Service(
          added_exceptions=set({
            datetime.date(2024, 3, 14),
          }),
          end=datetime.date(1, 1, 1),
          removed_exceptions=set({
          }),
          service_days=ServiceDays(
            friday=False,
            monday=False,
            saturday=False,
            sunday=False,
            thursday=False,
            tuesday=False,
            wednesday=False,
          ),
          service_id='PiDaySpecial',
          start=datetime.date(1, 1, 1),
        ),
        'Regular': Service(
          added_exceptions=set({
          }),
          end=datetime.date(2024, 12, 31),
          removed_exceptions=set({
            datetime.date(2024, 12, 25),
          }),
          service_days=ServiceDays(
            friday=True,
            monday=True,
            saturday=False,
            sunday=False,
            thursday=True,
            tuesday=True,
            wednesday=True,
          ),
          service_id='Regular',
          start=datetime.date(2024, 1, 1),
        ),
        'Weekday': Service(
          added_exceptions=set({
          }),
          end=datetime.date(2024, 12, 31),
          removed_exceptions=set({
          }),
          service_days=ServiceDays(
            friday=True,
            monday=True,
            saturday=False,
            sunday=False,
            thursday=True,
            tuesday=True,
            wednesday=True,
          ),
          service_id='Weekday',
          start=datetime.date(2024, 1, 1),
        ),
        'Weekend': Service(
          added_exceptions=set({
          }),
          end=datetime.date(2024, 12, 31),
          removed_exceptions=set({
          }),
          service_days=ServiceDays(
            friday=False,
            monday=False,
            saturday=True,
            sunday=True,
            thursday=False,
            tuesday=False,
            wednesday=False,
          ),
          service_id='Weekend',
          start=datetime.date(2024, 1, 1),
        ),
      }),
    }),
    'route_info_ds': dict({
      'route_infos': dict({
        'A': RouteInfo(
          agency_id='TEST AGENCY',
          color='111111',
          desc='Runs a test route A, all times',
          id='A',
          long_name='A Stop Time Test Route',
          short_name='A',
          text_color='',
          type=<RouteType.SUBWAY: 1>,
          url='https://gtfs.example.com/a',
        ),
        'X': RouteInfo(
          agency_id='TEST AGENCY',
          color='EE352E',
          desc='Runs a test route X, all times',
          id='X',
          long_name='X Test Route',
          short_name='X',
          text_color='FFFFFF',
          type=<RouteType.SUBWAY: 1>,
          url='https://gtfs.example.com/x',
        ),
        'Y': RouteInfo(
          agency_id='TEST AGENCY',
          color='00933C',
          desc='Runs a test route Y, all times',
          id='Y',
          long_name='Y Test Route',
          short_name='Y',
          text_color='FFFFFF',
          type=<RouteType.SUBWAY: 1>,
          url='https://gtfs.example.com/y',
        ),
        'Z': RouteInfo(
          agency_id='TEST AGENCY',
          color='B933AD',
          desc='Runs a test route Z, sometimes',
          id='Z',
          long_name='Z Test Route',
          short_name='Z',
          text_color='',
          type=<RouteType.SUBWAY: 1>,
          url='https://gtfs.example.com/z',
        ),
      }),
    }),
    'station_stop_info_ds': dict({
      'station_stop_infos': dict({
        '101': 101: Test Station Main St, lat: 40.60, long: -74.5,
        '101N': 101N: Test Station Main St, lat: 40.60, long: -74.5, parent: 101,
        '101S': 101S: Test Station Main St, lat: 40.60, long: -74.5, parent: 101,
        '102': 102: Test Station 10 St, lat: 40.70, long: -74.6,
        '102N': 102N: Test Station 10 St, lat: 40.70, long: -74.6, parent: 102,
        '102S': 102S: Test Station 10 St, lat: 40.70, long: -74.6, parent: 102,
        '103': 103: Test Station 20 St, lat: 40.70, long: -74.6,
        '103N': 103N: Test Station 20 St, lat: 40.70, long: -74.6, parent: 103,
        '103S': 103S: Test Station 20 St, lat: 40.70, long: -74.6, parent: 103,
      }),
    }),
    'stop_times_ds': dict({
      'stop_times': dict({
      }),
      'stops_filter': set({
      }),
    }),
    'trip_info_ds': dict({
      'trip_infos': dict({
        '123_X..N01R': 123_X..N01R: X to Northbound X,
        '123_Y..N02R': 123_Y..N02R: Y to Northbound Y,
        '123_Z..N03R': 123_Z..N03R: Z to Northbound Z,
        '321_X..S01R': 321_X..S01R: X to Southbound X,
        '321_Y..S02R': 321_Y..S02R: Y to Southbound Y,
        '321_Z..S03R': 321_Z..S03R: Z to Southbound Z,
        '456_X..N04R': 456_X..N04R: X to Northbound X,
        '456_Y..N05R': 456_Y..N05R: Y to Northbound Y,
        '456_Z..N06R': 456_Z..N06R: Z to Northbound Z,
        '654_X..S04R': 654_X..S04R: X to Southbound X,
        '654_Y..S05R': 654_Y..S05R: Y to Southbound Y,
        '654_Z..S06R': 654_Z..S06R: Z to Southbound Z,
        'STOP_TIME_TRIP': STOP_TIME_TRIP: A to Southbound A,
        'bad': bad:  to Eastbound ?,
      }),
    }),
  })
# ---
# name: test_async_build_schedule_add_data_later
  dict({
    'calendar': dict({
      'services': dict({
        'NoFridayService': Service(
          added_exceptions=set({
          }),
          end=datetime.date(1, 1, 1),
          removed_exceptions=set({
            datetime.date(2024, 3, 1),
          }),
          service_days=ServiceDays(
            friday=False,
            monday=False,
            saturday=False,
            sunday=False,
            thursday=False,
            tuesday=False,
            wednesday=False,
          ),
          service_id='NoFridayService',
          start=datetime.date(1, 1, 1),
        ),
        'PiDaySpecial': Service(
          added_exceptions=set({
            datetime.date(2024, 3, 14),
          }),
          end=datetime.date(1, 1, 1),
          removed_exceptions=set({
          }),
          service_days=ServiceDays(
            friday=False,
            monday=False,
            saturday=False,
            sunday=False,
            thursday=False,
            tuesday=False,
            wednesday=False,
          ),
          service_id='PiDaySpecial',
          start=datetime.date(1, 1, 1),
        ),
        'Regular': Service(
          added_exceptions=set({
          }),
          end=datetime.date(2024, 12, 31),
          removed_exceptions=set({
            datetime.date(2024, 12, 25),
          }),
          service_days=ServiceDays(
            friday=True,
            monday=True,
            saturday=False,
            sunday=False,
            thursday=True,
            tuesday=True,
            wednesday=True,
          ),
          service_id='Regular',
          start=datetime.date(2024, 1, 1),
        ),
        'Weekday': Service(
          added_exceptions=set({
          }),
          end=datetime.date(2024, 12, 31),
          removed_exceptions=set({
          }),
          service_days=ServiceDays(
            friday=True,
            monday=True,
            saturday=False,
            sunday=False,
            thursday=True,
            tuesday=True,
            wednesday=True,
          ),
          service_id='Weekday',
          start=datetime.date(2024, 1, 1),
        ),
        'Weekend': Service(
          added_exceptions=set({
          }),
          end=datetime.date(2024, 12, 31),
          removed_exceptions=set({
          }),
          service_days=ServiceDays(
            friday=False,
            monday=False,
            saturday=True,
            sunday=True,
            thursday=False,
            tuesday=False,
            wednesday=False,
          ),
          service_id='Weekend',
          start=datetime.date(2024, 1, 1),
        ),
      }),
    }),
    'route_info_ds': dict({
      'route_infos': dict({
        'A': RouteInfo(
          agency_id='TEST AGENCY',
          color='111111',
          desc='Runs a test route A, all times',
          id='A',
          long_name='A Stop Time Test Route',
          short_name='A',
          text_color='',
          type=<RouteType.SUBWAY: 1>,
          url='https://gtfs.example.com/a',
        ),
        'X': RouteInfo(
          agency_id='TEST AGENCY',
          color='EE352E',
          desc='Runs a test route X, all times',
          id='X',
          long_name='X Test Route',
          short_name='X',
          text_color='FFFFFF',
          type=<RouteType.SUBWAY: 1>,
          url='https://gtfs.example.com/x',
        ),
        'Y': RouteInfo(
          agency_id='TEST AGENCY',
          color='00933C',
          desc='Runs a test route Y, all times',
          id='Y',
          long_name='Y Test Route',
          short_name='Y',
          text_color='FFFFFF',
          type=<RouteType.SUBWAY: 1>,
          url='https://gtfs.example.com/y',
        ),
        'Z': RouteInfo(
          agency_id='TEST AGENCY',
          color='B933AD',
          desc='Runs a test route Z, sometimes',
          id='Z',
          long_name='Z Test Route',
          short_name='Z',
          text_color='',
          type=<RouteType.SUBWAY: 1>,
          url='https://gtfs.example.com/z',
        ),
      }),
    }),
    'station_stop_info_ds': dict({
      'station_stop_infos': dict({
        '101': 101: Test Station Main St, lat: 40.60, long: -74.5,
        '101N': 101N: Test Station Main St, lat: 40.60, long: -74.5, parent: 101,
        '101S': 101S: Test Station Main St, lat: 40.60, long: -74.5, parent: 101,
        '102': 102: Test Station 10 St, lat: 40.70, long: -74.6,
        '102N': 102N: Test Station 10 St, lat: 40.70, long: -74.6, parent: 102,
        '102S': 102S: Test Station 10 St, lat: 40.70, long: -74.6, parent: 102,
        '103': 103: Test Station 20 St, lat: 40.70, long: -74.6,
        '103N': 103N: Test Station 20 St, lat: 40.70, long: -74.6, parent: 103,
        '103S': 103S: Test Station 20 St, lat: 40.70, long: -74.6, parent: 103,
        '201': 201: Test Station Last St, lat: 40.60, long: -74.5, parent: 201,
        '201N': 201N: Test Station Last St, lat: 40.60, long: -74.5, parent: 201,
      }),
    }),
    'stop_times_ds': dict({
      'stop_times': dict({
      }),
      'stops_filter': set({
      }),
    }),
    'trip_info_ds': dict({
      'trip_infos': dict({
        '123_X..N01R': 123_X..N01R: X to Northbound X,
        '123_Y..N02R': 123_Y..N02R: Y to Northbound Y,
        '123_Z..N03R': 123_Z..N03R: Z to Northbound Z,
        '321_X..S01R': 321_X..S01R: X to Southbound X,
        '321_Y..S02R': 321_Y..S02R: Y to Southbound Y,
        '321_Z..S03R': 321_Z..S03R: Z to Southbound Z,
        '456_X..N04R': 456_X..N04R: X to Northbound X,
        '456_Y..N05R': 456_Y..N05R: Y to Northbound Y,
        '456_Z..N06R': 456_Z..N06R: Z to Northbound Z,
        '654_X..S04R': 654_X..S04R: X to Southbound X,
        '654_Y..S05R': 654_Y..S05R: Y to Southbound Y,
        '654_Z..S06R': 654_Z..S06R: Z to Southbound Z,
        '789_X..N24R': 789_X..N24R: X to Northbound Special X,
        '789_X..S24R': 789_X..S24R: X to Southbound Special X,
        '987_X..N21R': 987_X..N21R: X to Northbound Special X,
        '987_X..S21R': 987_X..S21R: X to Southbound Special X,
        'STOP_TIME_TRIP': STOP_TIME_TRIP: A to Southbound A,
        'bad': bad:  to Eastbound ?,
      }),
    }),
  })
# ---
# name: test_stop_time_filtering
  dict({
    'calendar': dict({
      'services': dict({
        'NoFridayService': Service(
          added_exceptions=set({
          }),
          end=datetime.date(1, 1, 1),
          removed_exceptions=set({
            datetime.date(2024, 3, 1),
          }),
          service_days=ServiceDays(
            friday=False,
            monday=False,
            saturday=False,
            sunday=False,
            thursday=False,
            tuesday=False,
            wednesday=False,
          ),
          service_id='NoFridayService',
          start=datetime.date(1, 1, 1),
        ),
        'PiDaySpecial': Service(
          added_exceptions=set({
            datetime.date(2024, 3, 14),
          }),
          end=datetime.date(1, 1, 1),
          removed_exceptions=set({
          }),
          service_days=ServiceDays(
            friday=False,
            monday=False,
            saturday=False,
            sunday=False,
            thursday=False,
            tuesday=False,
            wednesday=False,
          ),
          service_id='PiDaySpecial',
          start=datetime.date(1, 1, 1),
        ),
        'Regular': Service(
          added_exceptions=set({
          }),
          end=datetime.date(2024, 12, 31),
          removed_exceptions=set({
            datetime.date(2024, 12, 25),
          }),
          service_days=ServiceDays(
            friday=True,
            monday=True,
            saturday=False,
            sunday=False,
            thursday=True,
            tuesday=True,
            wednesday=True,
          ),
          service_id='Regular',
          start=datetime.date(2024, 1, 1),
        ),
        'Weekday': Service(
          added_exceptions=set({
          }),
          end=datetime.date(2024, 12, 31),
          removed_exceptions=set({
          }),
          service_days=ServiceDays(
            friday=True,
            monday=True,
            saturday=False,
            sunday=False,
            thursday=True,
            tuesday=True,
            wednesday=True,
          ),
          service_id='Weekday',
          start=datetime.date(2024, 1, 1),
        ),
        'Weekend': Service(
          added_exceptions=set({
          }),
          end=datetime.date(2024, 12, 31),
          removed_exceptions=set({
          }),
          service_days=ServiceDays(
            friday=False,
            monday=False,
            saturday=True,
            sunday=True,
            thursday=False,
            tuesday=False,
            wednesday=False,
          ),
          service_id='Weekend',
          start=datetime.date(2024, 1, 1),
        ),
      }),
    }),
    'route_info_ds': dict({
      'route_infos': dict({
        'A': RouteInfo(
          agency_id='TEST AGENCY',
          color='111111',
          desc='Runs a test route A, all times',
          id='A',
          long_name='A Stop Time Test Route',
          short_name='A',
          text_color='',
          type=<RouteType.SUBWAY: 1>,
          url='https://gtfs.example.com/a',
        ),
        'X': RouteInfo(
          agency_id='TEST AGENCY',
          color='EE352E',
          desc='Runs a test route X, all times',
          id='X',
          long_name='X Test Route',
          short_name='X',
          text_color='FFFFFF',
          type=<RouteType.SUBWAY: 1>,
          url='https://gtfs.example.com/x',
        ),
        'Y': RouteInfo(
          agency_id='TEST AGENCY',
          color='00933C',
          desc='Runs a test route Y, all times',
          id='Y',
          long_name='Y Test Route',
          short_name='Y',
          text_color='FFFFFF',
          type=<RouteType.SUBWAY: 1>,
          url='https://gtfs.example.com/y',
        ),
        'Z': RouteInfo(
          agency_id='TEST AGENCY',
          color='B933AD',
          desc='Runs a test route Z, sometimes',
          id='Z',
          long_name='Z Test Route',
          short_name='Z',
          text_color='',
          type=<RouteType.SUBWAY: 1>,
          url='https://gtfs.example.com/z',
        ),
      }),
    }),
    'station_stop_info_ds': dict({
      'station_stop_infos': dict({
        '101': 101: Test Station Main St, lat: 40.60, long: -74.5,
        '101N': 101N: Test Station Main St, lat: 40.60, long: -74.5, parent: 101,
        '101S': 101S: Test Station Main St, lat: 40.60, long: -74.5, parent: 101,
        '102': 102: Test Station 10 St, lat: 40.70, long: -74.6,
        '102N': 102N: Test Station 10 St, lat: 40.70, long: -74.6, parent: 102,
        '102S': 102S: Test Station 10 St, lat: 40.70, long: -74.6, parent: 102,
        '103': 103: Test Station 20 St, lat: 40.70, long: -74.6,
        '103N': 103N: Test Station 20 St, lat: 40.70, long: -74.6, parent: 103,
        '103S': 103S: Test Station 20 St, lat: 40.70, long: -74.6, parent: 103,
      }),
    }),
    'stop_times_ds': dict({
      'stop_times': dict({
        'STOP_TIME_TRIP': dict({
          1: dict({
          }),
        }),
      }),
      'stops_filter': set({
        '101N',
      }),
    }),
    'trip_info_ds': dict({
      'trip_infos': dict({
        '123_X..N01R': 123_X..N01R: X to Northbound X,
        '123_Y..N02R': 123_Y..N02R: Y to Northbound Y,
        '123_Z..N03R': 123_Z..N03R: Z to Northbound Z,
        '321_X..S01R': 321_X..S01R: X to Southbound X,
        '321_Y..S02R': 321_Y..S02R: Y to Southbound Y,
        '321_Z..S03R': 321_Z..S03R: Z to Southbound Z,
        '456_X..N04R': 456_X..N04R: X to Northbound X,
        '456_Y..N05R': 456_Y..N05R: Y to Northbound Y,
        '456_Z..N06R': 456_Z..N06R: Z to Northbound Z,
        '654_X..S04R': 654_X..S04R: X to Southbound X,
        '654_Y..S05R': 654_Y..S05R: Y to Southbound Y,
        '654_Z..S06R': 654_Z..S06R: Z to Southbound Z,
        'STOP_TIME_TRIP': STOP_TIME_TRIP: A to Southbound A,
        'bad': bad:  to Eastbound ?,
      }),
    }),
  })
# ---
